datalad/core/local/resulthooks.py

```diff
@@ -0,0 +1,177 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Interface utility functions
+
+"""
+
+__docformat__ = 'restructuredtext'
+
+import logging
+import json
+
+lgr = logging.getLogger('datalad.core.local.hooks')
+
+
+def get_hooks_from_config(cfg):
+    """Parse out hook definitions given a ConfigManager instance
+
+    Returns
+    -------
+    dict
+      where keys are hook names/labels, and each value is a dict with
+      three keys: 'cmd' contains the name of the to-be-executed DataLad
+      command; 'args' has a JSON-encoded string with a dict of keyword
+      arguments for the command (format()-language based placeholders
+      can be present); 'match' holds a JSON-encoded string representing
+      a dict with key/value pairs that need to match a result in order
+      for a hook to be triggered.
+    """
+    hooks = {}
+    for h in (k for k in cfg.keys()
+              if k.startswith('datalad.result-hook.')
+              and k.endswith('.match')):
+        proc = cfg.get('{}.proc'.format(h[:-6]), None)
+        if not proc:
+            lgr.warning(
+                'Incomplete result hook configuration %s in %s' % (
+                    h[:-6], cfg))
+            continue
+        sep = proc.index(' ')
```

Comment:
Would there be a downside here of being more lenient and splitting on the first whitespace (`proc.split(maxsplit=1)`)?  Solely from a readability point, I'd find it nice to get rid of some of the indexing below.

---

Given that the `proc` and `match` values are coming from the user, it be good to show a bit more helpful message if they aren't valid.  For example, the `.index()` call above would fail with

    [ERROR  ] substring not found [resulthooks.py:get_hooks_from_config:45] (ValueError) 

if the value didn't have a space.  The `json.loads()` is another case (and perhaps more likely when users are trying to put valid json into a git config value). And misformatted arguments would make it through to `run_hook` before causing issues, so that's yet another spot.

Given that these hooks are secondary to the main command, perhaps these failures should give a warning and continue on with the rest of result processing.

Reply:
> Would there be a downside here of being more lenient and splitting on the first whitespace (`proc.split(maxsplit=1)`)? Solely from a readability point, I'd find it nice to get rid of some of the indexing below.

Oh yes! Thanks for reminding me that we are in PY3 land now!! Done.

I also added the ability to have argument-less call specifications, and a smoke test for them.

> Given that the `proc` and `match` values are coming from the user, it be good to show a bit more helpful message if they aren't valid. For example, the `.index()` call above would fail with
> 
> ```
> [ERROR  ] substring not found [resulthooks.py:get_hooks_from_config:45] (ValueError) 
> ```
> 
> if the value didn't have a space. The `json.loads()` is another case (and perhaps more likely when users are trying to put valid json into a git config value). And misformatted arguments would make it through to `run_hook` before causing issues, so that's yet another spot.
> 
> Given that these hooks are secondary to the main command, perhaps these failures should give a warning and continue on with the rest of result processing.

A broken match spec would now yield such warning:

```
WARNING: Invalid match specification in datalad.result-hook.annoy.match-json: {"type":["in", ["file"]],"action":"get",status":"notneeded"} [Expecting property name enclosed in double quotes: line 1 column 41 (char 40) [decoder.py:raw_decode:353]], hook will be skipped
```

Likewise, a broken call spec given something like this:

```
WARNING: Invalid argument specification for hook annoy (after parameter substitutions): {"cmd":"touch" /tmp/datalad_temp_test_basicsnscytin3/file1_annoyed","dataset":"/tmp/datalad_temp_test_basicsnscytin3","explicit":true} [Expecting ',' delimiter: line 1 column 16 (char 15) [decoder.py:raw_decode:353]], hook will be skipped
```